### DISCUSSION TITLE
Cannot configure MONAILabel Reviewer

### DISCUSSION THREAD

**Discussion by neerajamaha** [Project-MONAI/MONAILabel#1290] [Q&A]
Hello,

I am trying to set up MONAI Label Reviewer to review some Xray exams. Currently, I have converted all the datasets to NIFTI format and have the server running on my GPU. However, when I try to run MONAI Label Reviewer, My Slicer GUI continues to crash. I am using Slicer version 5.2,1 and MONAI Label version 0.6.0. 

I have also tried to run a previous version of Slicer (4.13.0) and MONAI Label (0.4.1), and I am able to get to the MONAI Label Reviewer module running on Slicer. However, when I connect to the server, I am not able to see any of the Data Evaluation parameters and I am also not able to load any exams (screenshot below).

I was wondering if MONAI Label Reviewer is dependent on the images running on the Orthanc DICOM server? If not, I was hoping I could get some assistance into setting up Reviewer mode so that I am able to load and review exams.

![image](https://user-images.githubusercontent.com/98112236/218120127-baee7647-3496-4230-bf8f-b85e67ad2047.png)

Thank you!

**Comment by jlvahldiek** [comment]
Thank you for raising this issue. I can confirm that the Reviewer is not fully working at the moment and is causing Slicer to crash. I will hopefully be able to look at it soon and hopefully find a quick fix.

Let's add @DaoMaoDuc  to the discussion.

**Comment by DaoMaoDuc** [comment]
Hi @neerajamaha 

Thanks a lot for raising the issue. 

Before you push the button "connect", could you please open the "Python Interactor" (which you can find the drop down menu of "View"). That will open a "Terminal"/"Console".
Afterwards, push the button "connect". You will see logs within that terminal.  Could you provide me these logs please?

Thanks a lot.

**Reply by neerajamaha** [reply]
Hi @DaoMaoDuc

Thank you for your reply. Here is the screenshot you've requested. 

![image](https://user-images.githubusercontent.com/98112236/218473791-dea24117-c140-4521-9b58-30f1fd15e099.png)
![image](https://user-images.githubusercontent.com/98112236/218473810-607c7591-f1c6-4821-bc8e-660263762ccf.png)

I noticed that there is a loadDicomAndSegmentation module being used to load the images and labels. As you can see by my data, I am loading in NIFTI files. Just for some context, I've converted the original DICOM files to NIFTI and then used the MONAI Label workflow to annotate the xray exams. I'd like to now implement the MONAI Label Reviewer mode so that the radiologist supervising the project can review them.

Please let me know if you need anymore information.

Thanks so much for your help!

**Reply by jlvahldiek** [reply]
Thank you for using the MONAILabelReviewer. This is exactly the purpose for which we developed and contributed it back then. Unfortunately, we had no resources to develop it further and to adapt it to the frequent changes of MONAILabel. That is why it no longer works today.

I have already found a few small things that I will contribute soon and then the Reviewer will hopefully run again.

Nevertheless, we contributed the Reviewer at that time in the hope of more contributions from the community. So please feel free to improve the Reviewer.

Certainly some design features need to be reworked and the naming of many functions is unfortunate. Nevertheless, we had the Reviewer as a valuable tool in use back then.

**Reply by chezhia** [reply]
Thanks @jlvahldiek @DaoMaoDuc for your responses. I work with @neerajamaha and would be willing to contribute to get MONAILabelReviewer working again as this is a pressing need in many of our projects. Kindly let us know if you want to setup a meeting.

**Reply by lassoan** [reply]
I am happy to help this work with any Slicer related questions. The best is if you can write to the Slicer forum (https://discourse.slicer.org) and mention me (@lassoan), but you can ask questions in this repository, too.

**Comment by jlvahldiek** [comment]
lets move this discussion to #1300

**Comment by jlvahldiek** [comment]
@neerajamaha, MONAILabelReviewer extension was fixed. It will take some time until the changes will be populated to Slicer's extension manager. Please clone latest MONAILabel repo and re-install MONAILabelReviewer plugin from there and try if that works for your use case.

Happy to chat about MONAILabelReviewer and further improvements, @chezhia.

**Reply by lassoan** [reply]
> It will take some time until the changes will be populated to Slicer's extension manager

You can get the updated extension for Slicer-5.2.1 by updating MONAILabel extension using the extensions manager tomorrow (2023-02-16) after 10am EST. The fix is also included in Slicer Preview Releases downloaded tomorrow or later.

**Reply by chezhia** [reply]
Thanks for such a quick turnaround @lassoan  and @jlvahldiek. We’ll try this update tomorrow and hopefully we can gain some experience and contribute for future improvements.

**Comment by neerajamaha** [comment]
Hi @lassoan @jlvahldiek @DaoMaoDuc 
Apologies for my delay in getting back to you about MONAI Label Reviewer. It seems that pulling the newest changes has resolved the crash issue. Thank you very much for turning that around so quickly.

I was able to load one exam and review the label but after submitting the edited label, I am not able to load in a new exam. When I restart slicer and relaunch Reviewer, it pulls in the same exam that I have already reviewed, but with some loading failures as well. I was wondering if I need to set any additional parameters to subsequently load exam and segmentation files? Here are the logs from the Python Console within Slicer. 

![image](https://user-images.githubusercontent.com/98112236/222743277-380281c5-4806-436d-99ac-e7b01d1294e2.png)

@chezhia and I can work on a solution but I first wanted to see if you all had seen this behavior before.

**Reply by DaoMaoDuc** [reply]
Hi everyone.
Thanks a lot for reporting that issue.
I have not seen such an behavior before.

Could you please provide us with the log messages before the Traceback of the Exception?

**Reply by jlvahldiek** [reply]
I could reproduce this issue. It originates from hard-coding the label's file extension and setting its value to `.nrrd`:
https://github.com/Project-MONAI/MONAILabel/blob/5d2ae1fbb770040ef1e428b8e7b0312c6bbead75/plugins/slicer/MONAILabelReviewer/MONAILabelReviewerLib/MonaiServerREST.py#L155
https://github.com/Project-MONAI/MONAILabel/blob/5d2ae1fbb770040ef1e428b8e7b0312c6bbead75/plugins/slicer/MONAILabelReviewer/MONAILabelReviewer.py#L1441

Fixing these two occasions may be easy but produces other issues...

Debugging MONAILabelReviewer is very time consuming due to nested code. MONAILabelReviewer was developed during a project in which we used a single DICOM file (image) with a .seg.nrrd label file. We did not test MONAILabelReviewer with other file formats. It was commited in its state of a prototype.

The tool is certainly needed by many other groups that use MONAILabel. I would like to strongly advocate to completely re-implement MONAILabelReviewer and use the complete API client of the MONAILabel Slicer extension. At the moment MONAILabelReviewer is more a stand-alone Slicer extension with a lot of code duplications/similar implementations of the MONAILabel Slicer extension.
